### ISSUE TITLE
[RFC] Generic autocomplete form type for resources (optional)

### ISSUE THREAD

**Issue by pjedrzejewski** [Sylius/Sylius#3884]
Hello again!

As mentioned in my other RFC about forms - https://github.com/Sylius/Sylius/issues/3816, we want to have a generic automomplete form type for any resource. I checked existing autocomplete bundles, ideas and possible solutions and would like to introduce our own ResourceAutocompleteType.

Goal:

```
<?php

// ...

class XyzType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        // ...
        $builder->add('product', 'sylius_product_autocomplete'); // For Sylius stuff.
        $builder->add('<any field>', '<application_name>_<resource_name>_autocomplete'); // For custom resources.
    }
}
```

Why not existing solutions:

- Some of them force us to use a concrete UI;
- Same with using a concrete API;
- Most if not all are ORM only;
- To maintain consistency with other form types;
- I have a simple idea :D

Proposed solution:

- ResourceAutocompleteType, which will take the MetadaInterface to generate appropriate name for the form type;
- It will have two fields internally:
    - ``$builder->add('input', 'text');``
    - ``$builder->add('identifier', '<application_name>_<resource_name>_to_hidden_identifier');``
- It will use a transformer to transform and reverseTransform ``array('input' => X, 'identifier' => Y) to just Y and vice versa. Rest will be handled by ResourceToHiddenIdentifierType. :)

Let's get it working like this and next step will be to attach an API to populate Select2 (in case of Sylius) with choices.

The form type would also take routing configuration to pass a proper url to the FormView for use in JS.

Thoughts?

**Comment by peteward** [comment]
Yes, a very good idea.

We have implemented one of these (badly) on our site so I'm sure you can do a much better job.

For us it's mandatory as any form that has a select box for something like... ProductVariant crashes the page for memory and time limit. 

Our frontend/admin implementation was to use Chosen (https://harvesthq.github.io/chosen/) which is a really good solution, although we didn't get as far as implementing multiple choices (we didn't have the use-case for it to start with, but it might be required).

![screen shot 2016-01-15 at 08 45 41](https://cloud.githubusercontent.com/assets/7068515/12349076/b523e102-bb64-11e5-8ebf-591968aa5562.png)

**Comment by tristanbes** [comment]
I'm sorry @peteward but I must disagree with you. Chosen and Select2 are not great libraries. But there is no great alternative I know (well I can think of one but it's [built for reactjs](https://github.com/JedWatson/react-select).

Nethertheless, select2 (v4.x) is more maintained than chosen. (see commit activiy for both repo)

**Comment by peteward** [comment]
OK yeah the code isn't great quality, but the functionality that it gives would work well here.

**Comment by tristanbes** [comment]
> we didn't get as far as implementing multiple choices
A great candidate would be promotions applied to certain products/categories

**Comment by peteward** [comment]
:+1:

**Comment by tristanbes** [comment]
The tricky part is to provide something flexible enough to handle:

- Multiple select
- Single selection
- Remote loading (to prevent loading huge set of data) with json/jsonp
- No remote loading (for fields we know that won't have a huge collection of data)
- Depandant select

The transformer would be generic right ? with ressource in mind ? to avoid having a transformer per ressource. How much you want it decoupled from UI/Widget (pretty hard I think).

[Genemu form bundle](https://github.com/genemu/GenemuFormBundle) (now inactive) offered such things.

**Comment by pjedrzejewski** [comment]
Initial functionality implemented in #3951. We added generic hidden type and SemanticUI has pretty neat functionality for different autocomplete variations. We will be integrating more in the new admin. :)
